render-app/src/main/java/org/janelia/alignment/match/OrderedCanvasIdPair.java

```diff
@@ -55,6 +57,47 @@ public OrderedCanvasIdPair(final CanvasId oneId,
         this.absoluteDeltaZ = deltaZ == null ? null : Math.abs(deltaZ);
     }
 
+    /**
+     * @param  oneTileBounds      identifiers and bounds for one canvas identifier.
+     * @param  anotherTileBounds  identifiers and bounds for another canvas identifier.
+     *
+     * @return an ordered pair where each CanvasId includes relative position information
+     *         based upon the specified tile bounds.
+     *         Assumes that the bounds are for the same layer so the pair's deltaZ is set to null.
+     *
+     * @throws IllegalArgumentException
+     *   if both tile identifiers are the same.
+     */
+    public static OrderedCanvasIdPair withRelativePositions(final TileBounds oneTileBounds,
+                                                            final TileBounds anotherTileBounds)
+            throws IllegalArgumentException {
```

Comment:
It seems to me that there is some redundancy or even misplaced responsibility in the classes used here:
* An `OrderedCanvasIdPair` can be constructed by using `TileBounds` as well as `CanvasId`, which was surprising to me because I didn't expect `TileBounds` to carry an id. If that is to be desired, I would suggest letting `TileBounds` store a `CanvasId` with an appropriate getter.
* `CanvasId` carries a relative geometric information in xy-direction, whereas `OrderedCanvasIdPair` contains a relative position in z-direction. Both geometric informations seem to be optional. I would suggest encapsulating relative geometric information into one class and only store a reference to an instance in `OrderedCanvasIdPair`.

Keep in mind: those are just suggestions. There might be better ways to improve the issues mentioned, or there might be arguments against changing the classes at all. Furthermore, this doesn't need to be addressed within this PR.

Reply:
Yeah - I see what you mean.  In retrospect, I should have looked into merging the concepts when they evolved.  Here is how/why (I think) I created the spaghetti: `TileBounds` were created to couple tile context (tileId and sectionId) with bounds info to support lighter-weight APIs where transformation details were not needed.  The `MontageRelativePosition` was added to `CanvasId` when we started clipping tiles to optimize point match derivation for same z-layer tiles.  The `absoluteDeltaZ` was added to `OrderedCanvasIdPair` when I introduced multi-stage matching which needed to toss too-distant pairs in later stages.

I think you have the right idea to encapsulate all relative positional information for a pair into a single class.  I'm adding a TODO for this since I don't have time to clean this up right now.